runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,567 @@
+    // most normal termination waiting mode; couldn't be done implicitly inside Bank::freeze() -> () to return
+    // the result and timing in some way to higher-layer subsystems;
+    TerminatedToFreeze,
+    DroppedFromBankForks,
+    // scheduler will be restarted without being returned to pool in order to reuse it immediately.
+    ReinitializedForRecentBlockhash,
+}
+
+pub type DefaultInstalledSchedulerBox = Box<dyn InstalledScheduler<DefaultScheduleExecutionArg>>;
+
+#[derive(Clone, Debug)]
+pub struct SchedulingContext {
+    mode: SchedulingMode,
+    // Intentionally not using Weak<Bank> for performance reasons
+    bank: Arc<Bank>,
+}
+
+impl WithSchedulingMode for SchedulingContext {
+    fn mode(&self) -> SchedulingMode {
+        self.mode
+    }
+}
+
+impl SchedulingContext {
+    pub fn new(mode: SchedulingMode, bank: Arc<Bank>) -> Self {
+        Self { mode, bank }
+    }
+
+    pub fn bank(&self) -> &Arc<Bank> {
+        &self.bank
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.bank().slot()
+    }
+
+    pub fn log_prefix(context: Option<&Self>, scheduler_id: SchedulerId) -> String {
+        const BITS_PER_HEX_DIGIT: usize = 4;
+
+        format!(
+            "sch_{:0width$x}{}: ",
+            scheduler_id,
+            context
+                .map(|c| format!("(slot:{}/mode:{:?})", c.slot(), c.mode()))
+                .unwrap_or_else(|| "(?)".into()),
+            width = SchedulerId::BITS as usize / BITS_PER_HEX_DIGIT,
+        )
+    }
+}
+
+// tiny wrapper to ensure to call wait_for_termination() via ::drop() inside
+// BankForks::set_root()'s pruning.
+pub struct BankWithScheduler {
+    inner: Arc<BankWithSchedulerInner>,
+}
+
+#[derive(Debug)]
+pub struct BankWithSchedulerInner {
+    bank: Arc<Bank>,
+    scheduler: InstalledSchedulerRwLock,
+}
+pub type InstalledSchedulerRwLock = RwLock<Option<DefaultInstalledSchedulerBox>>;
+
+#[allow(clippy::declare_interior_mutable_const)]
+pub const NO_INSTALLED_SCHEDULER_RW_LOCK: InstalledSchedulerRwLock = RwLock::new(None);
+
+impl BankWithScheduler {
+    pub(crate) fn new(bank: Arc<Bank>, scheduler: Option<DefaultInstalledSchedulerBox>) -> Self {
+        if let Some(bank_in_context) = scheduler.as_ref().and_then(|scheduler| scheduler.context())
+        {
+            assert_eq!(bank.slot(), bank_in_context.slot());
+        }
+
+        Self {
+            inner: Arc::new(BankWithSchedulerInner {
+                bank,
+                scheduler: RwLock::new(scheduler),
+            }),
+        }
+    }
+
+    #[cfg(any(test, feature = "test-in-workspace"))]
+    pub fn new_for_test(bank: Arc<Bank>, scheduler: Option<DefaultInstalledSchedulerBox>) -> Self {
+        Self::new(bank, scheduler)
+    }
+
+    pub fn new_without_scheduler(bank: Arc<Bank>) -> Self {
+        Self::new(bank, None)
+    }
+
+    pub fn clone_with_scheduler(&self) -> BankWithScheduler {
+        BankWithScheduler {
+            inner: self.inner.clone(),
+        }
+    }
```

Comment:
Should this be `impl Clone for BankWithScheduler`?

Reply:
That would shadow `<Arc<Bank> as Clone>::clone()`, given the existence of `impl Deref for BankWithScheduler`... so i can't do it. maybe i need to comment about this deliberate subtlety.

Reply:
I had a suspicious that `Deref` abuse will bring all that bad interactions that the C++ implicit conversions are bringing.
Here is a one indication of that :P

Maybe remove the `Deref`?  I think Rust style is to avoid implicit conversions except for a very few clearly defined cases that are almost part of the language in the first place.
Sorry if this suggestion goes completely against what you were doing.
Just making sure you consider that :)

Reply:
> I had a suspicious that `Deref` abuse will bring all that bad interactions that the C++ implicit conversions are bringing.
> Here is a one indication of that :P

yeah, i was reluctant as well for the use of `Deref` because of these unwanted effects...

> Maybe remove the Deref? I think Rust style is to avoid implicit conversions except for a very few clearly defined cases that are almost part of the language in the first place.

... that said, i still think removing `Deref` isn't desired, considering yet another rather large code churns by doing so..

also, it seems there are other funny `Deref`s already (just run `$ git grep -A5 "impl.*Deref"`). so, I'm not the first.. xD

> Sorry if this suggestion goes completely against what you were doing.
> Just making sure you consider it as a possibility :)

thanks for raising voices. I'll go as-is with some good (hopefully) doc: 3843793058dc49c9416e347afc26c605f7e576a7